From 9012f56d8d73d0898c607509f361585175408601 Mon Sep 17 00:00:00 2001
From: Thu Nguyen <thu@os.amperecomputing.com>
Date: Sun, 20 Jul 2025 09:12:40 +0000
Subject: [PATCH] redfish_host_interface: Support GetFingerPrint

Support `GetFingerPrint` Ipmi command as section `8.1.1 Get manager
certificate fingerprint (NetFn 2Ch, Command 01h)` in DSP0270 V1.3.1.
The code will use OpenSSL APIs to get 32 bytes Finger Print from file
`/etc/ssl/certs/https/server.pem`.

Tested:
```
ipmitool raw 0x2c 0x1 0x52 0x1
 xx xx xx xx xx xx xx xx xx xx xx xx xx xx xx xx
 xx xx xx xx xx xx xx xx xx xx xx xx xx xx xx xx
 xx xx
```

Upstream-Status: Pending

Change-Id: I749af7e5f6d6e5adfbe3b88a332a29d5f9d5f092
Signed-off-by: Thu Nguyen <thu@os.amperecomputing.com>
---
 include/ipmid/api-types.hpp |   3 +
 meson.build                 |   9 ++
 meson.options               |   8 ++
 redfishhostinterface.cpp    | 214 ++++++++++++++++++++++++++++++++++++
 redfishhostinterface.hpp    |  55 +++++++++
 5 files changed, 289 insertions(+)
 create mode 100644 redfishhostinterface.cpp
 create mode 100644 redfishhostinterface.hpp

diff --git a/include/ipmid/api-types.hpp b/include/ipmid/api-types.hpp
index 52c68ee..a88a8df 100644
--- a/include/ipmid/api-types.hpp
+++ b/include/ipmid/api-types.hpp
@@ -31,6 +31,9 @@ constexpr Group groupPICMG = 0x00;
 constexpr Group groupDMTG = 0x01;
 constexpr Group groupSSI = 0x02;
 constexpr Group groupVSO = 0x03;
+#ifdef REDFISH_HOST_INTERFACE
+constexpr Group groupRedfish = 0x52;
+#endif
 #ifdef ARM_SBMR_SUPPORT
 constexpr Group groupSBMR = 0xAE;
 #endif
diff --git a/meson.build b/meson.build
index 3dff44c..6cc9a97 100644
--- a/meson.build
+++ b/meson.build
@@ -105,6 +105,7 @@ feature_map = {
     'dynamic-storages-only'      : '-DFEATURE_DYNAMIC_STORAGES_ONLY',
     'dynamic-sensors-static-fru' : '-DFEATURE_DYNAMIC_STATIC_FRU_ID',
     'arm-sbmr'                   : '-DARM_SBMR_SUPPORT',
+    'redfish-host-interface'     : '-DREDFISH_HOST_INTERFACE',
 }
 
 foreach option_key, option_value : feature_map
@@ -256,6 +257,7 @@ ipmid_pre = [
     crypto,
     ipmid_dep,
     channellayer_dep,
+    userlayer_dep,
 ]
 
 transportoem_src = []
@@ -280,6 +282,11 @@ if get_option('arm-sbmr').allowed()
     arm_sbmr_cmds_src = ['sbmrhandler.cpp']
 endif
 
+redfish_host_interface_src = []
+if get_option('redfish-host-interface').enabled()
+    redfish_host_interface_src = ['redfishhostinterface.cpp']
+endif
+
 libipmi20_src = [
     'app/channel.cpp',
     'app/watchdog.cpp',
@@ -299,11 +306,13 @@ libipmi20_src = [
     'read_fru_data.cpp',
     'sensordatahandler.cpp',
     'user_channel/channelcommands.cpp',
+    'user_channel/user_layer.cpp',
     generated_src,
     transportoem_src,
     storage_cmds_src,
     openpower_cmds_src,
     arm_sbmr_cmds_src,
+    redfish_host_interface_src,
     conf_h,
 ]
 
diff --git a/meson.options b/meson.options
index cdbab9c..dfb23df 100644
--- a/meson.options
+++ b/meson.options
@@ -257,6 +257,14 @@ option(
     description: 'Support Arm SBMR specific functions',
 )
 
+# redfish-host-interface specific functionality.
+option(
+    'redfish-host-interface',
+    type: 'feature',
+    value: 'enabled',
+    description: 'Support redfish host interface specific functions',
+)
+
 #######################################################################
 # Transport options
 #######################################################################
diff --git a/redfishhostinterface.cpp b/redfishhostinterface.cpp
new file mode 100644
index 0000000..00d19d0
--- /dev/null
+++ b/redfishhostinterface.cpp
@@ -0,0 +1,214 @@
+/*
+ * Copyright (c) 2025 Ampere Computing LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "redfishhostinterface.hpp"
+
+#include <openssl/bio.h>
+#include <openssl/ec.h>
+#include <openssl/err.h>
+#include <openssl/evp.h>
+#include <openssl/pem.h>
+#include <openssl/rand.h>
+#include <openssl/x509.h>
+
+#include <ipmid/api.hpp>
+#include <ipmid/types.hpp>
+#include <ipmid/utils.hpp>
+#include <phosphor-logging/elog-errors.hpp>
+#include <phosphor-logging/lg2.hpp>
+#include <user_channel/user_layer.hpp>
+#include <user_channel/user_mgmt.hpp>
+
+using namespace phosphor::logging;
+
+static constexpr const char* credBootstrappingInf =
+    "xyz.openbmc_project.HostInterface.CredentialBootstrapping";
+static constexpr const char* credBootstrapEnabledProp = "Enabled";
+
+bool checkRHIAllowedMediumType(const uint8_t& mediumType)
+{
+    if (mediumType ==
+            static_cast<uint8_t>(ipmi::EChannelMediumType::smbusV20) ||
+        mediumType ==
+            static_cast<uint8_t>(ipmi::EChannelMediumType::systemInterface) ||
+        mediumType == static_cast<uint8_t>(ipmi::EChannelMediumType::oem))
+    {
+        return true;
+    }
+
+    return false;
+}
+
+std::shared_ptr<X509> loadCert(const std::string& filePath)
+{
+    std::shared_ptr<BIO> certFileBio{BIO_new_file(filePath.c_str(), "rb"),
+                                     BIO_free};
+    if (certFileBio == nullptr)
+    {
+        lg2::error("Error occurred during BIO_new_file call, path {PATH}",
+                   "PATH", filePath);
+        return nullptr;
+    }
+
+    X509* cert = nullptr;
+    if (PEM_read_bio_X509(certFileBio.get(), &cert, nullptr, nullptr) ==
+        nullptr)
+    {
+        lg2::error("Error occurred during PEM_read_bio_X509 call, path {PATH}",
+                   "PATH", filePath);
+        return nullptr;
+    }
+
+    return std::shared_ptr<X509>(cert, X509_free);
+}
+
+/** @brief implements the get CredentialBootstrapping's Enabled property
+ *
+ *  @returns 1 - enabled, 0 - disabled, std::nullopt if error
+ */
+std::optional<bool> getCredentialBootstrapEnabledProperty(
+    ipmi::Context::ptr ctx)
+{
+    std::string service;
+    bool enabled = false;
+
+    try
+    {
+        ipmi::Value bootstrapProperty = ipmi::getDbusProperty(
+            *ctx->bus, ipmi::userMgrService, ipmi::userObjBasePath,
+            credBootstrappingInf, credBootstrapEnabledProp);
+        enabled = std::get<bool>(bootstrapProperty);
+    }
+    catch (const std::exception& e)
+    {
+        lg2::error(
+            "Can't get value of {PROP} property in the {INF} interface at path {PATH} error {ERROR}",
+            "PROP", credBootstrapEnabledProp, "INF", credBootstrappingInf,
+            "PATH", ipmi::userObjBasePath, "ERROR", e);
+        return std::nullopt;
+    }
+
+    return enabled;
+}
+
+/** @brief implements the Get manager certificate fingerprint command
+ *
+ *  @param[in] ctx - shared_ptr to an IPMI context struct
+ *  @param[in] certNum - Certificate number
+ *
+ *  @returns IPMI completion code plus response data for
+ *   - the Get manager certificate fingerprint command
+ */
+ipmi::RspType<uint8_t,              // hash type
+              std::vector<uint8_t>> // hash data
+    getFingerprint(ipmi::Context::ptr ctx, uint8_t certNum)
+{
+    ipmi::ChannelInfo chInfo;
+    const std::string certFile = "/etc/ssl/certs/https/server.pem";
+    uint8_t cnBuffer[EVP_MAX_MD_SIZE];
+    unsigned int cnBufferLen;
+
+    if (certNum != 0x1)
+    {
+        lg2::error("Invalid certNum {CER}.", "CER", certNum);
+        return ipmi::responseCertsNumberInvalid();
+    }
+
+    if (ipmi::getChannelInfo(ctx->channel, chInfo) != ipmi::ccSuccess)
+    {
+        lg2::error("Failed to get Channel Info, channel={CHANNEL}", "CHANNEL",
+                   ctx->channel);
+        return ipmi::responseUnspecifiedError();
+    }
+
+    if (!checkRHIAllowedMediumType(chInfo.mediumType))
+    {
+        lg2::error("Error - Medium interface not supported, medium={TYPE}",
+                   "TYPE", chInfo.mediumType);
+        return ipmi::responseCommandNotAvailable();
+    }
+
+    /* Get Enabled property within the CredentialBootstrapping property of the
+     * host interface. */
+    auto enabledState = getCredentialBootstrapEnabledProperty(ctx);
+    if (!enabledState || (*enabledState == false))
+    {
+        lg2::error("CredentialBootstrapping is disabled.");
+        return ipmi::responseCmdDisabled();
+    }
+
+    auto cert = loadCert(certFile);
+    if (cert == nullptr)
+    {
+        lg2::error("Failed to read cert");
+        return ipmi::responseUnspecifiedError();
+    }
+
+    if (!X509_digest(cert.get(), EVP_sha256(), cnBuffer, &cnBufferLen))
+    {
+        lg2::error("Failed to get finger print.");
+        return ipmi::responseUnspecifiedError();
+    }
+
+    if (cnBufferLen != ipmi::maxFingerPrintLength)
+    {
+        lg2::error("Failed to get finger print.");
+        return ipmi::responseUnspecifiedError();
+    }
+    constexpr uint8_t sha256Type = 0x01;
+    std::vector<uint8_t> hashData{cnBuffer, cnBuffer + cnBufferLen};
+
+    return ipmi::responseSuccess(sha256Type, hashData);
+}
+
+ipmi::Cc filterRHICommands(ipmi::message::Request::ptr request)
+{
+    if (request->ctx->netFn != ipmi::netFnGroup ||
+        request->ctx->group != ipmi::groupRedfish)
+    {
+        // Skip if not group SBMR
+        return ipmi::ccSuccess;
+    }
+
+    ipmi::ChannelInfo chInfo;
+    if (ipmi::getChannelInfo(request->ctx->channel, chInfo) != ipmi::ccSuccess)
+    {
+        lg2::error("Failed to get Channel Info, channel={CHANNEL}", "CHANNEL",
+                   request->ctx->channel);
+        return ipmi::ccUnspecifiedError;
+    }
+
+    if (!checkRHIAllowedMediumType(chInfo.mediumType))
+    {
+        lg2::error("Error - Medium interface not supported, medium={TYPE}",
+                   "TYPE", chInfo.mediumType);
+        return ipmi::ccCommandNotAvailable;
+    }
+
+    return ipmi::ccSuccess;
+}
+
+void registerNetfnRedfishHostInterfaceFunctions() __attribute__((constructor));
+void registerNetfnRedfishHostInterfaceFunctions()
+{
+    ipmi::registerGroupHandler(ipmi::prioOpenBmcBase, ipmi::groupRedfish,
+                               ipmi::bootStrap::cmdGetMngCertFingerprint,
+                               ipmi::Privilege::User, getFingerprint);
+    ipmi::registerFilter(ipmi::prioOpenBmcBase,
+                         [](ipmi::message::Request::ptr request) {
+                             return filterRHICommands(request);
+                         });
+}
diff --git a/redfishhostinterface.hpp b/redfishhostinterface.hpp
new file mode 100644
index 0000000..6e5ade9
--- /dev/null
+++ b/redfishhostinterface.hpp
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2025 Ampere Computing LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <ipmid/api.hpp>
+#include <ipmid/types.hpp>
+
+#include <cstddef>
+#include <cstdint>
+
+namespace ipmi
+{
+/* BootStrap Account commands */
+/* Completion code */
+constexpr Cc ccCredsBootstrapDisabled = 0x80;
+constexpr Cc ccCertificateNumberInvalid = 0xCB;
+
+/* Credential bootstrapping control option */
+constexpr uint8_t keepCredBootstrapEnabled = 0xa5;
+
+/* 32 bytes of the fingerprint */
+static constexpr size_t maxFingerPrintLength = 32;
+
+namespace bootStrap
+{
+constexpr Cmd cmdGetMngCertFingerprint = 0x01;
+constexpr Cmd cmdGetBootstrapAccoutCre = 0x02;
+} // namespace bootStrap
+
+/* helper functions for the various bootStrap command error response types */
+static inline auto responseCmdDisabled()
+{
+    return response(ccCredsBootstrapDisabled);
+}
+
+static inline auto responseCertsNumberInvalid()
+{
+    return response(ccCertificateNumberInvalid);
+}
+
+} // namespace ipmi
-- 
2.34.1

